### PR TITLE
91 - PSM Routing

### PR DESCRIPTION
**Pull request by Preston-Harrison**
Routes are now weighted and convert returns the highest weighted quote and swap. 
If a token is pegged on the PSM, and has the highest route weighting, tokens will be swapped using the PSM.

### CODE REVIEW

File: src/config/hlp.ts [hunk 1]
```diff
@@ -79,7 +82,8 @@ export const HLP_CONTRACTS: NetworkMap<HlpContracts | undefined> = {
     Router: "0xa3815EB3D81f45634E33fBC67CD0C1177Ba3A131",
     Reader: "0xCb7AEB7f471D1c19C78E3cd578ee5Ff0788278B6",
     HLP: "0xB666b08609b2E69A8ba51AA720770053AeC0d2d3",
-    HlpManagerRouter: "0x3ecB21eABEF68a237862E8A003807eE4Fa47509b"
+    HlpManagerRouter: "0x3ecB21eABEF68a237862E8A003807eE4Fa47509b",
+    HPSM: "0xa2b81201F92b2F3081e9e2900Cf01942e0BCCeD3"
```
Comment (rellfy) [inline src/config/hlp.ts:86]: This is not an HLP contract. We should place it under `config/index.ts`, `config.protocol.arbitrum,protocol`

File: src/config/hlp.ts [hunk 2]
```diff
@@ -53,6 +56,7 @@ export type HlpContracts = {
   HlpManager: string;
   HLP: string;
   HlpManagerRouter: string;
+  HPSM: string;
```
Comment (rellfy) [inline src/config/hlp.ts:59]: This is not an HLP contract. It needs to be placed with the other config.

File: src/config/hlp.ts [hunk 3]
```diff
@@ -3,7 +3,10 @@ import config from ".";
 import { Network, NetworkMap } from "../types/network";
 
 /** Swap function gas limit for Arbitrum. */
-export const HLP_SWAP_GAS_LIMIT = "1500000";
+export const HLP_SWAP_GAS_LIMIT = 1_500_000;
+
+/** Peg stability gas limit */
+export const PSM_GAS_LIMIT = 1_000_000;
```
Comment (rellfy) [inline src/config/hlp.ts:9]: Should not be here.

**Review by rellfy — CHANGES_REQUESTED** [review]
Please remove PSM code from HLP files

**Review by xcjlo985 — COMMENTED** [review]
In general it is hard to marry up the code and the tests. Usually for a file `mycode.ts` you'd have a matching `mycode.test.ts`. Can we improve this?

File: src/config/hlp.ts (hunk 1 above)
Comment (Preston-Harrison) [reply src/config/hlp.ts]: done

File: src/config/hlp.ts (hunk 2 above)
Comment (Preston-Harrison) [reply src/config/hlp.ts]: done

File: src/config/hlp.ts (hunk 3 above)
Comment (Preston-Harrison) [reply src/config/hlp.ts]: done

**Comment by Preston-Harrison** [comment]
> In general it is hard to marry up the code and the tests. Usually for a file `mycode.ts` you'd have a matching `mycode.test.ts`. Can we improve this?

Agreed. I will make a PR into this branch that splits the tests up into their respective files

**Comment by Preston-Harrison** [comment]
I have merged the refactored tests into this branch

File: src/components/Convert/routes/hlpAddRemove.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    } else {
+      // if both tokens are not native
+      return hlpManager.populateTransaction.removeLiquidity(
+        toAddress,
+        BigNumber.from(sellAmount),
+        buyAmountWithTolerance,
+        connectedAccount
+      );
+    }
+  } else {
```
Comment (rellfy) [inline src/components/Convert/routes/hlpAddRemove.ts:142]: Can replace these by an `if (fromToken.symbol === "hLP" && toToken.isNative) { } else if (fromToken.symbol === "hLP") {}` to reduce indentation, and then get rid of the else in line 142.

File: src/components/Convert/routes/hlpSwap.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      { value: sellAmount }
+    );
+  } else {
+    return router.populateTransaction.swapTokensToETH(
+      [fromAddress, toAddress],
+      sellAmount,
+      buyAmountWithTolerance,
+      connectedAccount
+    );
+  }
```
Comment (rellfy) [inline src/components/Convert/routes/hlpSwap.ts:100]: Can replace `else if` by a new `if`, and remove the `else`.

File: src/components/Convert/routes/weth.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  const weth = getNativeWrappedToken(network)?.address;
+  if (!weth) throw new Error(`No WETH contract found for ${network}`);
+
+  if (fromToken.isNative && toToken.address === weth) {
+    return WETH__factory.connect(weth, signer).populateTransaction.deposit({
+      value: sellAmount
+    });
+  } else if (toToken.isNative && fromToken.address === weth) {
+    return WETH__factory.connect(weth, signer).populateTransaction.withdraw(sellAmount);
+  }
```
Comment (rellfy) [inline src/components/Convert/routes/weth.ts:43]: Can replace `else if` by a new `if`

File: src/config/index.ts [hunk 7]
```diff
@@ -65,6 +65,11 @@ export type Config = {
       nonStable: number;
     };
     tokenSymbolToStableType: { [key: string]: StableType };
+    gasEstimates: {
+      hPsm: number;
```
Comment (rellfy) [inline src/config/index.ts:69]: I don't think we need to capitalise the `P` here, so it matches the style of the `hlp` variable.  So camelCase would be `hpsm` or `somethingHpsm`. Can you please refactor these?

**Review by rellfy — CHANGES_REQUESTED** [review]

File: src/components/Convert/routes/hlpAddRemove.ts [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+      );
+
+    return {
+      allowanceTarget: hlpManagerAddress,
+      sellAmount: fromAmount.toString(),
+      buyAmount: buyAmount.toString(),
+      gas: config.convert.gasEstimates.hlp,
+      feeBasisPoints: feeBasisPoints.toNumber()
+    };
+  }
```
[the remark this answers is not in the document]
Comment (Preston-Harrison) [reply src/components/Convert/routes/hlpAddRemove.ts:87]: done

File: src/components/Convert/routes/hlpAddRemove.ts (hunk 4 above)
Comment (Preston-Harrison) [reply src/components/Convert/routes/hlpAddRemove.ts:142]: done

File: src/components/Convert/routes/hlpSwap.ts (hunk 5 above)
Comment (Preston-Harrison) [reply src/components/Convert/routes/hlpSwap.ts:100]: done

File: src/components/Convert/routes/weth.ts (hunk 6 above)
Comment (Preston-Harrison) [reply src/components/Convert/routes/weth.ts:43]: done

**Review by rellfy — APPROVED** [review]

File: src/config/index.ts (hunk 7 above)
Comment (Preston-Harrison) [reply src/config/index.ts]: Done

**Comment by Preston-Harrison** [comment]
@nabsta-eth plz re review

**Review by nabsta-eth — APPROVED** [review]

**Pull request merged**